source/lac/petsc_matrix_base.cc

```diff
@@ -89,14 +89,23 @@ namespace PETScWrappers
     AssertThrow(ierr == 0, ExcPETScError(ierr));
   }
 
+  void
+  MatrixBase::assign_petsc_matrix(Mat A)
+  {
+    AssertThrow(last_action == ::dealii::VectorOperation::unknown,
+                ExcMessage("Cannot assign a new Mat"));
+    PetscErrorCode ierr =
+      PetscObjectReference(reinterpret_cast<PetscObject>(A));
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    destroy_matrix(matrix);
+    matrix = A;
```

Comment:
You're not keeping track of ownership. That means that in the destructor of this class, you'll call `destroy_matrix()` on the so-assigned `matrix` (==A) object, but that then basically means that the current class takes over ownership.

Or does `PetscObjectReference` above increment a use counter, and then `destroy_matrix` really only does something when the reference count becomes zero? If the latter, please add a couple of lines of comment to the effect above.

Reply:
Yes, objects in PETSc are basically shared pointers.  `PetscObjectReference` increments reference count, XXXDestroy decrements it and, if refcount goes to zero, also destroys the inner data. There's no concept of unique ownership.

Reply:
I've used PETSc for nearly 20 years now and had never realized this :-)